mesonbuild/dependencies/base.py

```diff
@@ -498,31 +496,56 @@ def __init__(self, name, environment, kwargs, language=None):
-                if self.required:
-                    raise DependencyException('Pkg-config binary missing from cross file')
+        if not self.want_cross and environment.is_cross_build():
+            for_machine = MachineChoice.BUILD
+        else:
+            for_machine = MachineChoice.HOST
+
+        # Create a nested function for sake of early return
+        def search():
+            # Only search for the pkg-config for each machine the first time and
+            # store the result in the class definition
+            if PkgConfigDependency.class_pkgbin[for_machine] is None:
+                mlog.debug('Pkg-config binary for %s is not cached.' % for_machine)
+            else:
+                mlog.debug('Pkg-config binary for %s is cached.' % for_machine)
+                choice = PkgConfigDependency.class_pkgbin[for_machine]
+                assert choice is not None
+                return choice
+            # Lookup in cross or machine file.
+            bt = environment.binaries[for_machine]
+            potential_pkgpath = bt.lookup_entry('pkgconfig')
+            if potential_pkgpath is None:
+                mlog.debug('Pkg-config binary missing from cross or native file, or PKG_CONFIG undefined.')
             else:
-                potential_pkgbin = ExternalProgram.from_bin_list(
-                    environment.cross_info.config['binaries'], 'pkgconfig')
-                if potential_pkgbin.found():
-                    self.pkgbin = potential_pkgbin
+                mlog.debug('Pkg-config binary for %s specified from config file as %s.', for_machine, potential_pkgpath)
+                potential_pkgbin = ExternalProgram.from_entry('pkgconfig', potential_pkgpath)
+                if not potential_pkgbin.found():
+                    mlog.debug(
+                        'Pkg-config %s for machine %s specified at %s but not found.',
+                        potential_pkgbin.name, for_machine, potential_pkgbin.command)
                 else:
-                    mlog.debug('Cross pkg-config %s not found.' % potential_pkgbin.name)
-        # Only search for the native pkg-config the first time and
-        # store the result in the class definition
-        elif PkgConfigDependency.class_pkgbin is None:
-            self.pkgbin = self.check_pkgconfig()
-            PkgConfigDependency.class_pkgbin = self.pkgbin
-        else:
-            self.pkgbin = PkgConfigDependency.class_pkgbin
+                    return potential_pkgbin
+            # Fallback on hard-coded defaults.
+            if environment.machines.matches_build_machine(for_machine):
+                for potential_pkgpath in environment.default_pkgconfig:
+                    potential_pkgbin = self.check_pkgconfig(potential_pkgpath)
+                    if potential_pkgbin is None:
+                        mlog.debug(
+                             'default Pkg-config fallback %s for machine %s specified at %s but not found.',
```

Comment:
__[Flake8]__ <sub>posted by Sider</sub>

[E126] continuation line over-indented for hanging indent